api/odl.py

```diff
@@ -359,17 +411,183 @@ def fulfill(self, patron, pin, licensepool, internal_format):
             expires,
         )
 
+    def _update_hold_end_date(self, hold):
+        _db = Session.object_session(hold)
+        pool = hold.license_pool
+
+        # First make sure the hold position is up-to-date, since we'll
+        # need it to calculate the end date.
+        self._update_hold_position(hold)
+
+        default_loan_period = self.collection(_db).default_loan_period(
+            hold.patron.library
+        )
+        default_reservation_period = self.collection(_db).default_reservation_period
+
+        # If the hold is ready to check out and already has an end date,
+        # it doesn't need an update.
+        if hold.position == 0 and hold.end:
+            return
+
+        # If the patron is in the queue, we need to estimate when the book
+        # will be available for check out. We can do slightly better than the
+        # default calculation since we know when all current loans will expire,
+        # but we're still calculating the worst case.
+        elif hold.position > 0:
+            # The licenses will have to go through some number of cycles
+            # before one of them gets to this hold. This leavs out the first cycle -
+            # it's already started so we'll handle it separately.
+            cycles = (hold.position - pool.licenses_reserved - 1) / pool.licenses_owned
+
+            # Each of the owned licenses is currently either on loan or reserved.
+            # Figure out which license this hold will eventually get if every
+            # patron keeps their loans and holds for the maximum time.
+            copy_index = (hold.position - pool.licenses_reserved - 1)  % pool.licenses_owned
+
+            # Find the current loans and reserved holds for the licenses.
+            current_loans = _db.query(Loan).filter(Loan.license_pool_id==pool.id).order_by(Loan.start).all()
+            current_reservations = _db.query(Hold).filter(Hold.license_pool_id==pool.id).order_by(Hold.start).limit(pool.licenses_reserved).all()
+
+            # In the worse case, the first cycle ends when a current loan expires, or
+            # after a current reservation is checked out and then expires.
+            if len(current_loans) > copy_index:
+                next_cycle_start = current_loans[copy_index].end
+            else:
+                reservation = current_reservations[copy_index - len(current_loans)]
+                if reservation.end:
+                    next_cycle_start = reservation.end + datetime.timedelta(days=default_loan_period)
+                else:
+                    # TODO: If there's no reservation period and someone has a reserved hold, 
+                    # the maximum wait time is infinite.
+                    # Maybe reservation period should be required so this won't happen?
```

Comment:
I think it's fine to make reservation period required and set it to 3 days by default.